### PR TITLE
Exclude comment sections from workspace symbols by default

### PR DESCRIPTION
**Pull request by lionel-**
Branched from #865.
Addresses https://github.com/posit-dev/positron/issues/4886.

When a comment section like `# name ---` clashes with an actual symbol like `name <- function() {}`, whichever symbols comes first (usually the section) is included in the workspace index, and the other is ignored. Ideally we'd track both but for now we only track one identifier per file.

This causes sections to win over function definitions in cases like this:

```r
# my_section ----

my_section <- function() {}

my_section # Reference
```

If you then jump to definition on a reference to that function, Positron jumps to the section instead of the function.

* To fix this, we now allow functions and variables to overwrite sections in the indexer.

* In addition, we no longer emit sections as workspace symbols by default. This can be turned back on with a new setting `positron.r.workspaceSymbols.includeCommentSections`. This is consistent with the fix in https://github.com/quarto-dev/quarto/pull/755 where we no longer export markdown headers as workspace symbols by default, to avoid flooding workspace symbol quickpicks (`#` prefix in command palette) with section headers.

### QA Notes

- You should be able to command+quick on the `my_section` reference and be taken to the function definition rather than the section. This should be the case no matter the value of `positron.r.workspaceSymbols.includeCommentSections`.

- When `positron.r.workspaceSymbols.includeCommentSections` is set to `true`, you should see comment sections in the workspace symbol quickpick. Otherwise they shouldn't be included.

### CODE REVIEW

File: crates/ark/src/lsp/indexer.rs [hunk 1]
```diff
@@ -121,10 +121,17 @@ fn insert(path: &Path, entry: IndexEntry) -> anyhow::Result<()> {
...
+    // future we'll track every occurrences and their scopes but for now we
+    // only track the first definition of an object (in a way, its
     // declaration).
-    if !index.contains_key(&entry.key) {
+    if let Some(existing_entry) = index.get(&entry.key) {
+        // Give priority to non-section entries.
+        if matches!(existing_entry.data, IndexEntryData::Section { .. }) {
+            index.insert(entry.key.clone(), entry);
+        }
+        // Else, ignore.
```
Comment (DavisVaughan) [inline crates/ark/src/lsp/indexer.rs:143]: I think a test for this behavior would be nice

File: crates/ark/src/lsp/symbols.rs [hunk 2]
```diff
@@ -81,18 +84,21 @@ pub fn symbols(params: &WorkspaceSymbolParams) -> anyhow::Result<Vec<SymbolInfor
...
-                    kind: SymbolKind::STRING,
-                    location: Location {
-                        uri: Url::from_file_path(path).unwrap(),
-                        range: entry.range,
-                    },
-                    tags: None,
-                    deprecated: None,
-                    container_name: None,
-                });
+                if state.config.workspace_symbols.include_comment_sections {
```
Comment (DavisVaughan) [inline crates/ark/src/lsp/symbols.rs:87]: Would enjoy seeing a test for this (on and off) as well if possible

File: crates/ark/src/lsp/symbols.rs [hunk 3]
```diff
@@ -81,18 +84,21 @@ pub fn symbols(params: &WorkspaceSymbolParams) -> anyhow::Result<Vec<SymbolInfor
...
+                        tags: None,
+                        deprecated: None,
+                        container_name: None,
+                    });
+                }
             },
+
             IndexEntryData::Variable { name } => {
                 info.push(SymbolInformation {
                     name: name.clone(),
```
Comment (DavisVaughan) [inline crates/ark/src/lsp/symbols.rs]: Could also add a nice test in `goto_definition` for

```
# my_section ----

my_section <- function() {}
```

to make sure that `my_section` is chosen

I know I'm also asking for a similar test in `indexer` itself, but it feels like having a duplicate test in both places is a good idea (so if we ever see the `indexer` test and think "why is this here?", then we'd also have the goto-definition test to show the real reason for it)

**Review by DavisVaughan — APPROVED** [review]
Looks good to me, mostly just want some test coverage since this is a bit tricky

**Comment by lionel-** [comment]
We've now got the requested test coverage and a test-only RAII indexer guard to help clean up the index after tests have run.
